internal/pkg/ntp/ntp.go

```diff
@@ -0,0 +1,326 @@
+	timeServers    []string
+	lastSyncServer string
+
+	timeSyncNotified bool
+	timeSynced       chan struct{}
+
+	restartSyncCh chan struct{}
+	epochChangeCh chan struct{}
+
+	MinPoll, MaxPoll time.Duration
+
+	// these functions are overridden in tests for mocking support
+	CurrentTime CurrentTimeFunc
+	NTPQuery    QueryFunc
+	SetTime     SetTimeFunc
+	AdjustTime  AdjustTimeFunc
+}
+
+// NewSyncer creates new Syncer with default configuration.
+func NewSyncer(logger *log.Logger, timeServers []string) *Syncer {
+	syncer := &Syncer{
+		logger: logger,
+
+		timeServers: append([]string(nil), timeServers...),
+		timeSynced:  make(chan struct{}),
+
+		restartSyncCh: make(chan struct{}, 1),
+		epochChangeCh: make(chan struct{}, 1),
+
+		MinPoll: MinAllowablePoll,
+		MaxPoll: MaxAllowablePoll,
+
+		CurrentTime: time.Now,
+		NTPQuery:    ntp.Query,
+		SetTime:     syscall.Settimeofday,
+		AdjustTime:  timex.Adjtimex,
+	}
+
+	return syncer
+}
+
+// Synced returns a channel which is closed when time is in sync.
+func (syncer *Syncer) Synced() <-chan struct{} {
+	return syncer.timeSynced
+}
+
+// EpochChange returns a channel which receives a value each time jumps more than EpochLimit.
+func (syncer *Syncer) EpochChange() <-chan struct{} {
+	return syncer.epochChangeCh
+}
+
+func (syncer *Syncer) getTimeServers() []string {
+	syncer.timeServersMu.Lock()
+	defer syncer.timeServersMu.Unlock()
+
+	return syncer.timeServers
+}
+
+func (syncer *Syncer) getLastSyncServer() string {
+	syncer.timeServersMu.Lock()
+	defer syncer.timeServersMu.Unlock()
+
+	return syncer.lastSyncServer
+}
+
+func (syncer *Syncer) setLastSyncServer(lastSyncServer string) {
+	syncer.timeServersMu.Lock()
+	defer syncer.timeServersMu.Unlock()
+
+	syncer.lastSyncServer = lastSyncServer
+}
+
+// SetTimeServers sets the list of time servers to use.
+func (syncer *Syncer) SetTimeServers(timeServers []string) {
+	syncer.timeServersMu.Lock()
+	defer syncer.timeServersMu.Unlock()
+
+	if reflect.DeepEqual(timeServers, syncer.timeServers) {
+		return
+	}
+
+	syncer.timeServers = append([]string(nil), timeServers...)
+	syncer.lastSyncServer = ""
+
+	syncer.restartSync()
+}
+
+func (syncer *Syncer) restartSync() {
+	select {
+	case syncer.restartSyncCh <- struct{}{}:
+	default:
+	}
+}
+
+// Run runs the sync process.
+//
+// Run is usually run in a goroutine.
+// When context is canceled, sync process aborts.
+//
+//nolint: gocyclo
+func (syncer *Syncer) Run(ctx context.Context) {
+	RTCClockInitialize.Do(func() {
+		var err error
+
+		RTCClock, err = rtc.OpenRTC()
+		if err != nil {
+			syncer.logger.Printf("failure opening RTC, ignored: %s", err)
+		}
+	})
+
+	for {
+		lastSyncServer, resp, err := syncer.query(ctx)
+		if err != nil {
+			return
+		}
+
+		if resp != nil && resp.Validate() == nil {
+			if err = syncer.adjustTime(resp.ClockOffset, lastSyncServer); err != nil {
+				syncer.logger.Printf("error adjusting time: %s", err)
+			} else if !syncer.timeSyncNotified {
```

Comment:
Use `switch` too?

Reply:
this is a function call, with error handling, so it shouldn't be turned into `switch`

Reply:
I mean, may we do the same as https://github.com/talos-systems/talos/pull/3360#discussion_r602187161 to eliminate `else if`?

Reply:
```go
err = syncer.adjustTime(resp.ClockOffset, lastSyncServer)
switch err {
  case nil:
    if !syncer.timeSyncNotified { … }
  default:
    syncer.logger.Printf("error adjusting time: %s", err)
}
```

Reply:
I don't like `switch` but I will look into rewriting this